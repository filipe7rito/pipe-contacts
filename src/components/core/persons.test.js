import {
  render,
  wait,
  fireEvent,
  getByLabelText
} from '@testing-library/react';
import React from 'react';
import Persons from './persons';

afterEach(() => {
  jest.clearAllMocks();
});

jest.mock('../../services/contacts.js', () => {
  const response = {
    data: {
      success: true,
      data: [
        { id: 1, name: 'Judas' },
        { id: 2, name: 'Tony' }
      ],
      additional_data: {
        pagination: {
          start: 0,
          limit: 10,
          more_items_in_collection: false
        }
      }
    }
  };

  const findUser = string => {
    const filteredData = response.data.data.filter(user => {
      return user.name.toUpperCase().match(string.toUpperCase());
    });

    const filteredResponse = {
      data: {
        ...response.data,
        data: filteredData
      }
    };

    return Promise.resolve(filteredResponse);
  };

  return {
    getUsers: jest.fn(() => Promise.resolve(response)),
    findUser: findUser
  };
});

describe('<Persons>', () => {
  test('Render persons component', () => {
    const { asFragment, debug } = renderPersons();

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders a list of Persons', async () => {
    const { getByText } = renderPersons();
    await wait();
    expect(getByText(/judas/i)).toBeInTheDocument();
  });

  it('Should search by name , filtering the current list', async () => {
    const { queryByText, getByPlaceholderText, debug } = renderPersons();
    const spyOn = jest.spyOn(Persons.prototype, 'findUser');

    await wait();

    fireEvent.change(getByPlaceholderText('search user'), {
      target: {
        value: 'Judas'
      }
    });

    await wait();

    expect(queryByText('Tony')).not.toBeInTheDocument();
    expect(spyOn).toHaveBeenCalledWith('judas');
  });

  function renderPersons() {
    return render(<Persons></Persons>);
  }
});
