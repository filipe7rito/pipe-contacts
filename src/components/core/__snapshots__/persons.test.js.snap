// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Persons> Render persons component 1`] = `
<DocumentFragment>
  <div
    class="persons-container"
  >
    <div>
      <h5
        class="m-3"
      >
        People's List
      </h5>
      <hr />
      <div
        class="row toolbar"
      >
        <div
          class="col-10"
        >
          <div>
            <form>
              <div
                class="form-group row"
              >
                <div
                  class="col-sm-5"
                >
                  <input
                    class="form-control"
                    placeholder="search user"
                  />
                  <small
                    class="form-text text-muted m-2"
                  >
                    At least 2 characters long
                  </small>
                </div>
              </div>
            </form>
          </div>
        </div>
        <div
          class="col-2"
        >
          <button
            class="btn btn-primary add-user"
          >
            Create user
          </button>
        </div>
      </div>
    </div>
    <div
      class="table-container"
    >
      <table
        class="table table-borderless"
      >
        <thead />
        <tbody />
      </table>
    </div>
    <nav
      aria-label="Page navigation example"
    >
      <ul
        class="pagination"
      >
        <li
          class="page-item disabled"
        >
          <button
            aria-label="Previous"
            class="page-link m-1"
          >
            <span
              aria-hidden="true"
            >
              Previous
            </span>
          </button>
        </li>
        <li
          class="page-item disabled"
        >
          <button
            aria-label="Next"
            class="page-link m-1"
          >
            <span
              aria-hidden="true"
            >
              Next
            </span>
          </button>
        </li>
      </ul>
    </nav>
  </div>
</DocumentFragment>
`;
